crates/stackable-certs/src/cert_builder.rs

```diff
@@ -0,0 +1,330 @@
+use x509_cert::{
+    builder::{Builder, Profile},
+    der::{DecodePem, asn1::Ia5String},
+    ext::pkix::{ExtendedKeyUsage, SubjectAltName, name::GeneralName},
+    name::Name,
+    serial_number::SerialNumber,
+    spki::SubjectPublicKeyInfoOwned,
+    time::Validity,
+};
+
+use crate::{
+    CertificatePair,
+    ca::{CertificateAuthority, DEFAULT_CERTIFICATE_VALIDITY, PEM_LINE_ENDING},
+    keys::CertificateKeypair,
+};
+
+/// Defines all error variants which can occur when creating a certificate
+#[derive(Debug, Snafu)]
+pub enum CreateCertificateError<E>
+where
+    E: std::error::Error + 'static,
+{
+    #[snafu(display("failed to parse validity"))]
+    ParseValidity { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to parse \"{subject}\" as subject"))]
+    ParseSubject {
+        source: x509_cert::der::Error,
+        subject: String,
+    },
+
+    #[snafu(display("failed to create key pair"))]
+    CreateKeyPair { source: E },
+
+    #[snafu(display("failed to serialize public key as PEM"))]
+    SerializePublicKey { source: x509_cert::spki::Error },
+
+    #[snafu(display("failed to decode SPKI from PEM"))]
+    DecodeSpkiFromPem { source: x509_cert::der::Error },
+
+    #[snafu(display("failed to create certificate builder"))]
+    CreateCertificateBuilder { source: x509_cert::builder::Error },
+
+    #[snafu(display("failed to add certificate extension"))]
+    AddCertificateExtension { source: x509_cert::builder::Error },
+
+    #[snafu(display("The subject alternative DNS name \"{dns_name}\" is not a Ia5String"))]
+    SaDnsNameNotAIa5String {
+        dns_name: String,
+        source: x509_cert::der::Error,
+    },
+
+    #[snafu(display("failed to build certificate"))]
+    BuildCertificate { source: x509_cert::builder::Error },
+}
+
+/// This builder builds certificates of type [`CertificatePair`].
+///
+/// Example code to construct a certificate:
+///
+/// ```no_run
+/// use stackable_certs::{
+///     keys::ecdsa,
+///     ca::{CertificateAuthority, CertificateAuthorityBuilder},
+///     CertificateBuilder,
+/// };
+///
+/// let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+///     .build()
+///     .build_ca()
+///     .expect("failed to build CA");
+///
+/// let certificate = CertificateBuilder::builder()
+///     .subject("CN=trino-coordinator-default-0")
+///     .signed_by(&ca)
+///     .build()
+///     .build_certificate()
+///     .expect("failed to build certificate");
+/// ```
+#[derive(Builder)]
+pub struct CertificateBuilder<'a, KP>
+where
+    KP: CertificateKeypair,
+    <KP::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    /// Required subject of the certificate, usually starts with `CN=`.
+    subject: &'a str,
+
+    /// Optional list of subject alternative names (SAN) DNS entries,
+    /// that are added to the certificate.
+    #[builder(default)]
+    subject_alterative_dns_names: &'a [&'a str],
```

Comment:
suggestion: I would just go with

```suggestion
    /// Optional list of subject alternative names (SAN) DNS entries,
    /// that are added to the certificate.
    #[builder(default)]
    subject_alterative_names: &'a [&'a str],
```

Reply:
Also, there is no value in carrying this as `&str`. They need to be converted into owned values below anyway. The API should reflect that to not hide (possibly) unexpected cloning/allocation.

Reply:
Oh I misread, I though @nightkr requested to support IPs... I added support for that in 8f86a93 :see_no_evil: 

But as this is a builder I prefer to have a function to add DNs names and one to add IP addresses instead of passing a list of enums (which we first need to introduce).

Regarding the allocation, I see the point that this allocates. On the other hand I prefer the simple api, so that not every caller needs to deal with `der::asn1::ia5_string::allocation::Ia5String`

Reply:
> Oh I misread, I though @nightkr requested to support IPs... I added support for that in https://github.com/stackabletech/operator-rs/commit/8f86a935a164309ac5b7974045bf9cb623b46ca8 🙈

I was. (At least from the perspective of "if we want to use it for secret-op at some point".)